guardian/prosemirror-elements
# Add collaborative editing proof of concept to demo page

## What does this change?

Adds a collaborative editing PoC to the demo. This lets us see a few things that were mostly proved in our tests, but are nice to experience directly:

- collab semantics in rich text fields (e.g. non-last-write-wins – tweak the `setInterval` value in the code to see concurrent edits resolve themselves via manual edits. Perhaps we could let users tweak this themselves?)
- nested decorations rendered (user caret positions are represented as decorations, which appear in nested Prosemirror fields). This has lead to a bugfix (#51).

![2021-07-05 17 41 27](https://user-images.githubusercontent.com/7767575/124789448-33832380-df42-11eb-82ac-a38f16616b82.gif)

It doesn't include tests for this sort of thing, as I think they'd be a fair amount of effort. They'd also be testing the the collab code – which is very rough and ready but not the purpose of the library – rather than the core library code. If we can think of things that'd be useful to test via collab, that might be interesting.

## Known issues

It looks like decorations at the very edge of the field boundary sometimes end up in the wrong place. This could be a mapping problem, or some other issue.

## How to test

In the demo, click 'add editor' and have a play. Hey, click it again! Does everything (outside of 'known issues!) work as expected?
